### ISSUE TITLE
ReferenceError: Can't find variable: babelHelpers

### ISSUE THREAD

**Issue by lookfirst** [jspm/jspm-cli#633]
0.15.1

https://travis-ci.org/lookfirst/systemjs-seed/jobs/56162735
ReferenceError: Can't find variable: babelHelpers

I'm not sure whether to file this under SystemJS or jspm, but again, this was working with jspm 0.14.

The really odd bit is that I can't seem to reproduce this locally.

**Comment by guybedford** [comment]
I'd be very interested to hear how this is happening. Any luck reproducing?

**Comment by lookfirst** [comment]
Ok, cleared my jspm cache, node_modules and jspm_packages and did a npm update -g and was able to reproduce this locally. If you'd like to play along...

1. git clone https://github.com/lookfirst/systemjs-seed.git
1. npm install
1. npm test

**Comment by lookfirst** [comment]
I've got `externalHelpers: true` in my es6 build step. This produces code like this:

`var angular = babelHelpers.interopRequire(require("angular"));`

But, babelHelpers isn't defined anywhere global...

**Comment by lookfirst** [comment]
Adding `require('babel/external-helpers');` to the top of my first loaded file fixes the issue.

**Comment by lookfirst** [comment]
I think I need this to be a bit smarter... 

https://github.com/systemjs/systemjs/blob/master/lib/extension-es6.js#L66

**Comment by lookfirst** [comment]
Ok, this above commit fixes the case when running as a webapp, but doesn't solve the case of running from karma, which doesn't load the entire app. Basically, I really need systemjs to add external-helpers.js all the time. I must be missing something here...

**Comment by lookfirst** [comment]
Ok, that is an ugly fix for loading external-helpers via karma.

**Comment by guybedford** [comment]
@lookfirst can you explain why https://github.com/systemjs/systemjs/blob/master/lib/extension-es6.js#L66 isn't able to load babelHelpers for you?

**Comment by guybedford** [comment]
The only restriction seems to be that it only applies to modules of format "register".

**Comment by guybedford** [comment]
Ahh you're compiling to CommonJS...

**Comment by guybedford** [comment]
You probably want to compile to `modules: system` in Babel.

**Comment by lookfirst** [comment]
Tried that and there is all sorts of path related issues then. For some reason the paths to the js files aren't appended from the config.js at runtime. 

Seems like a pretty needless restriction.

This used to work.

**Comment by lookfirst** [comment]
When I change to `modules: system`, I get this in the devtools: http://take.ms/PhYRP

**Comment by guybedford** [comment]
Are you doing your Babel compile through Gulp? In this case you are on your own to ensure the helpers exist.

**Comment by lookfirst** [comment]
Yes. 

That seems a bit heavy handed, no? Why can't SystemJS detect that I'm using the helpers (either by looking at the code or through configuration) and load them?

**Comment by guybedford** [comment]
We don't want to be parsing every single CommonJS file on npm to see if it contains babelHelpers. Also if something is published to npm as CommonJS you want it to be installable through npm too, which will give the same issue. If this isn't the case, then I don't understand why you're publishing to npm as CommonJS.

**Comment by lookfirst** [comment]
I'm not publishing anywhere. This is in my systemjs-seed project.

**Comment by lookfirst** [comment]
I compile my app using babel, then I want to load it with systemjs. That seems like a pretty standard workflow.

**Comment by guybedford** [comment]
Ok, sorry I misunderstood. Make sure you use `modules=system` with anonymous output - `moduleIds = false`. That should completely be working and is the primary compile target for ES6.

**Comment by lookfirst** [comment]
Lets document this somewhere.

**Comment by lookfirst** [comment]
I am using both modules=system and moduleIds false. it still isn't working.

**Comment by guybedford** [comment]
I just ran a blank test case here, writing a file test.js:

```javascript
System.register([], function(_export) {
  return {
    setters: [],
    execute: function() {
      var angular = babelHelpers.interopRequire();
    }
  }
});
```

And then loading that with SystemJS. The above executes completely fine and doesn't give `babelHelpers` is undefined at all for me.

Can you perhaps let me know how the test case is different from your scenario?

**Comment by lookfirst** [comment]
Right now, my generated code looks more like this:

```
System.register(["angular", "oclazyload-systemjs-router", "app/routes.json!"], function (_export) {
	var angular, router, futureRoutes, appModule;
	return {
		setters: [function (_angular) {
			angular = _angular["default"];
		}, function (_oclazyloadSystemjsRouter) {
			router = _oclazyloadSystemjsRouter["default"];
		}, function (_appRoutesJson) {
			futureRoutes = _appRoutesJson["default"];
		}],
		execute: function () {
			// This is a hack because systemjs isn't loading this for us. why not?
			// https://github.com/jspm/jspm-cli/issues/633
			//require('babel/external-helpers');

			"use strict";

			appModule = angular.module("app", []);

			appModule.config(router(appModule, futureRoutes));

			appModule.config(["$locationProvider", "$httpProvider", "$urlRouterProvider", function ($locationProvider, $httpProvider, $urlRouterProvider) {
				$locationProvider.html5Mode({
					enabled: true,
					requireBase: false
				});
				$httpProvider.useApplyAsync(true);
				return $urlRouterProvider.otherwise("/todo");
			}]);

			angular.element(document).ready(function () {
				return angular.bootstrap(document.body, [appModule.name], {
					strictDi: true
				});
			});

			_export("default", appModule);
		}
	};
});
//# sourceMappingURL=../app/app.js.map
```

and I'm seeing this in the devtools (same screenshot as before): http://take.ms/PhYRP

**Comment by lookfirst** [comment]
That was after my only change was to switch to `modules=system`

**Comment by guybedford** [comment]
That particular code has no `babelHelpers` global being used at all?

**Comment by lookfirst** [comment]
I can put some in there if you like. :smile: There is other code referenced from that code that does need babelHelpers. That code doesn't load because of the switch to modules=system.

**Comment by guybedford** [comment]
Try replacing your main entry point code with:

```javascript
System.register([], function(_export) {
  return {
    setters: [],
    execute: function() {
      var angular = babelHelpers.interopRequire();
    }
  }
});
```

Does it load babelHelpers then?

**Comment by lookfirst** [comment]
Oh, I see your point now! :bulb: :+1: 

What you are saying is that when I declare modules=system, babelHelpers should no longer be added to the transpiled source. I missed a setting, so I was still getting it in there in some other files. I've fixed things and babelHelpers is nowhere in my transpiled source tree.

Now, the issue is more that SystemJS doesn't seem to be paying any attention to my config.js. It isn't finding the paths to the defined modules, like angular, anymore. It just hits things at root as shown in the screenshot.

**Comment by lookfirst** [comment]
Oh, I see what might be the problem. My config.js is also getting processed through Babel and being converted into a module=system.

```
System.register([], function (_export) {
  return {
    setters: [],
    execute: function () {
      "use strict";

      System.config({
        baseURL: ".",
        transpiler: "babel",
        babelOptions: {
          moduleIds: false,
          optional: ["runtime"]
        },
        paths: {
          "*": "*.js",
          "github:*": "../jspm_packages/github/*.js",
          "npm:*": "../jspm_packages/npm/*.js"
        },
        buildCSS: true,
        separateCSS: false
      });
...
```

**Comment by lookfirst** [comment]
Ok, yea, getting rid of that wrapping fixes it.

**Comment by lookfirst** [comment]
Going to close this as user error.

**Comment by guybedford** [comment]
Glad to hear it's sorted finally, and not a bug this side!

**Comment by zeeshanjan82** [comment]
I am getting following error when deploying the release. Actually I am using NodeWebkit and making a release using this seed project
Potentially unhandled rejection [2] ReferenceError: Error loading "app/login/login" at file:///C:/Users/ADMINI~1/AppData/Local/Temp/nw9104_8193/dist/app/login/login.js
babelHelpers is not defined

**Comment by guybedford** [comment]
@zeeshanjan82 what type of bundling operation are you doing here? SFX or standard bundling? I assume you've disabled the runtime modules transformer in Babel to get this. If using standard bundling, it is up to you to manually include babelHelpers - it is not emitted as a separate dependency. There is an issue to track this in https://github.com/systemjs/systemjs/issues/431, but it is not currently being prioritised for implementation.

The short answer is to include the babel runtime file separately with a script tag in the head of the page.

**Comment by zeeshanjan82** [comment]
Hi @guybedford  Thanks for the reply. I am using the following seed project: https://github.com/Swimlane/angular-systemjs-seed and also am using NodeKit to convert my angularjs app into .exe file. Everything is working fine and I see that the seed project has already imported the babel external helper file but I am not sure why do I get this error

**Comment by zeeshanjan82** [comment]
Potentially unhandled rejection [2] ReferenceError: Error loading "app/login/login" at file:///C:/Users/ADMINI~1/AppData/Local/Temp/nw9272_1565/dist/app/login/login.js
babelHelpers is not defined
    at file:///C:/Users/ADMINI~1/AppData/Local/Temp/nw9272_1565/dist/app/login/login.js:1:394
    at execute (file:///C:/Users/ADMINI~1/AppData/Local/Temp/nw9272_1565/dist/app/login/login.js:1:951)
    at m (file:///C:/Users/ADMINI~1/AppData/Local/Temp/nw9272_1565/dist/app/app.js:9:6421)
    at m (file:///C:/Users/ADMINI~1/AppData/Local/Temp/nw9272_1565/dist/app/app.js:9:6356)
    at Object.execute (file:///C:/Users/ADMINI~1/AppData/Local/Temp/nw9272_1565/dist/app/app.js:9:8290)
    at o (file:///C:/Users/ADMINI~1/AppData/Local/Temp/nw9272_1565/dist/app/app.js:7:18496)
    at p (file:///C:/Users/ADMINI~1/AppData/Local/Temp/nw9272_1565/dist/app/app.js:7:18868)
    at j (file:///C:/Users/ADMINI~1/AppData/Local/Temp/nw9272_1565/dist/app/app.js:7:16406)
    at k (file:///C:/Users/ADMINI~1/AppData/Local/Temp/nw9272_1565/dist/app/app.js:7:16791)
    at file:///C:/Users/ADMINI~1/AppData/Local/Temp/nw9272_1565/dist/app/app.js:7:15192

**Comment by guybedford** [comment]
You need to manually include `external-helpers.js` file from the Babel package which will define the `babelHelpers` global.

**Comment by zeeshanjan82** [comment]
Thanks @guybedford  you have always been helpful. I have now changed my approach and I realized the way I was trying to achieve the functionality or have the web application work with NodeWebkit was completely wrong and did not make any sense. Because when I serve the same web app folder from express.js everything works great without any need to modify things but when I try not to use express and instead serve the files directly from filesystem then the problem comes. I have finally decided to have the web server serve the files and the problem is solved.
